### PR TITLE
Add format_checkers module for json schema; explicit checking of array elements

### PR DESCRIPTION
**Pull request by GernotMaier**
**Review after #PR 1190 is merged to main**

The usage of jsonschema ensures that formats and values have the correct format.

Introduce here a custom format checker module to simplify the usages for simtools-related custom formats:

- moved simtools.data_model.metadata_model.astropy_unit into simtools.data_model.format_checkers (this allows to test that a unit is consistent with an astropy unit in a model parameter json / db entry).
- added `simtools.data_model.format_checkers.array_element`, which allows to check that an entry is consistent with the simtools array element naming (using `names.validate_array_element_name`)

Maintenance updates:

- decouple format checker from jsonschema draft `Draft7Validator` (not sure why we have used it from the beginning)

## Copilot summary (generated automatically)

This pull request introduces custom format checkers for JSON schema validation, refactors existing validation logic to use these new checkers, and updates relevant tests and schemas to accommodate these changes.

### Custom Format Checkers for JSON Schema Validation:
* [`simtools/data_model/format_checkers.py`](diffhunk://#diff-abfa9dc7adb421e96a3ae076286df886f293d97e372abce176bb80ed2624a770R1-R25): Added custom format checkers for `astropy_unit` and `array_element` to validate astropy units and array elements, respectively.

### Refactoring Code to Use New Format Checkers:
* [`simtools/data_model/metadata_model.py`](diffhunk://#diff-e29c478f2ac735868adcfbe13c79f350fae5123576a533180569bf407d6fee52L14-L33): Removed existing format checker functions and updated the `validate_schema` function to use the new format checkers from `format_checkers`. [[1]](diffhunk://#diff-e29c478f2ac735868adcfbe13c79f350fae5123576a533180569bf407d6fee52L14-L33) [[2]](diffhunk://#diff-e29c478f2ac735868adcfbe13c79f350fae5123576a533180569bf407d6fee52L54-R44)
* [`simtools/data_model/validate_data.py`](diffhunk://#diff-eecb8454a9066ab719b15e61c0de45617d89f765b7c97caf7c5cb9fa5f15c8e2R8-R15): Added a new method `_validate_data_dict_using_json_schema` to validate data dictionaries using JSON schemas and the new format checkers. Updated `_validate_data_dict` to use this new method. [[1]](diffhunk://#diff-eecb8454a9066ab719b15e61c0de45617d89f765b7c97caf7c5cb9fa5f15c8e2R8-R15) [[2]](diffhunk://#diff-eecb8454a9066ab719b15e61c0de45617d89f765b7c97caf7c5cb9fa5f15c8e2L145-R149) [[3]](diffhunk://#diff-eecb8454a9066ab719b15e61c0de45617d89f765b7c97caf7c5cb9fa5f15c8e2R165-R185)

### Schema Updates:
* [`simtools/schemas/model_parameter_and_data_schema.metaschema.yml`](diffhunk://#diff-be3cf0fa965ee04a73de99ddf922c3313f7800fbaf76497b2c5e5e71e5e05832R101-R105): Added a `json_schema` property to the schema definitions.
* [`simtools/schemas/model_parameters/array_layouts.schema.yml`](diffhunk://#diff-f1c8a0bd3084f045e93a77202c32c6ba70593f36c6090dcd0dc2c73cacd3a377R13-R28): Updated the schema to include a `json_schema` for array layout definitions.

### Test Updates:
* [`tests/unit_tests/data_model/test_format_checkers.py`](diffhunk://#diff-c10e3f5cebac591f500342720bc1bbcfec1391600b355243546b1da18a02faf5R1-R17): Added new tests for the custom format checkers.
* [`tests/unit_tests/data_model/test_metadata_model.py`](diffhunk://#diff-61d8a0df29affe4790476b8ba9da60255b4b9a62ecd8239c195facb56403f53dL12-L16): Removed tests for the old format checker functions.
* [`tests/unit_tests/data_model/test_validate_data.py`](diffhunk://#diff-c1c68a857731db82f8d071d52024afdc53f85476788c9fe711f3979d97abd3a3R9): Updated tests to mock the `_prepare_model_parameter` method and added new tests for `_validate_data_dict_using_json_schema`. [[1]](diffhunk://#diff-c1c68a857731db82f8d071d52024afdc53f85476788c9fe711f3979d97abd3a3R9) [[2]](diffhunk://#diff-c1c68a857731db82f8d071d52024afdc53f85476788c9fe711f3979d97abd3a3L98-R99) [[3]](diffhunk://#diff-c1c68a857731db82f8d071d52024afdc53f85476788c9fe711f3979d97abd3a3R116-R123) [[4]](diffhunk://#diff-c1c68a857731db82f8d071d52024afdc53f85476788c9fe711f3979d97abd3a3R745-R773)

### CODE REVIEW

File: simtools/data_model/format_checkers.py
Comment (orelgueta) [inline simtools/data_model/format_checkers.py:11]: Does it actually raise a ValueError if you except it and return a string?

File: simtools/data_model/validate_data.py
Comment (orelgueta) [inline simtools/data_model/validate_data.py:147]: Codecov complains this line is not covered by tests.

File: simtools/data_model/format_checkers.py
Comment (orelgueta) [inline simtools/data_model/format_checkers.py:21]: If I understood correctly, it would actually raise `ValidationError` and not `ValueError`, no?

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (orelgueta) [inline tests/unit_tests/data_model/test_format_checkers.py:10]: How about testing also the case in which it fails?

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (orelgueta) [inline tests/unit_tests/data_model/test_format_checkers.py:16]: Oh, I see that indeed it raises a `ValueError`. So when would it raise `ValidationError`? Or is this test not complete?

File: simtools/schemas/model_parameter_and_data_schema.metaschema.yml
Comment (orelgueta) [inline simtools/schemas/model_parameter_and_data_schema.metaschema.yml:102]: Previously we used draft 7 and here it is draft 6. Does it matter?

**Review by orelgueta — COMMENTED** [review]
Some questions more than anything, but I do not approve yet just to make sure everything is OK before we merge.

File: simtools/data_model/format_checkers.py
Comment (GernotMaier) [reply simtools/data_model/format_checkers.py:11]: Thanks! I've copied this function from existing code in metadata_model without looking at it. Fixed this now.

File: simtools/data_model/validate_data.py
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:147]: Yes, I consciously didn't add it as I either would have to add a complex example with data and jsonschema or define at least 7 mocker patches. I don't want to do the latter and as this is kind of a trivial line I think it can be such.

(and I know that this is actually a hint that this function is too complex; suprised that sonar is not complaining)

File: simtools/data_model/format_checkers.py
Comment (GernotMaier) [reply simtools/data_model/format_checkers.py:21]: Yes, it raises ValueError. Motivation is to keep to standard python Exceptions and not to introduce the complex json schema exceptions

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (GernotMaier) [reply tests/unit_tests/data_model/test_format_checkers.py:10]: Added two tests for this.

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (GernotMaier) [reply tests/unit_tests/data_model/test_format_checkers.py:16]: Never. Why do you expect a ValidationError ? And this would be a `jsonschema.exceptions.ValidationError`

File: simtools/schemas/model_parameter_and_data_schema.metaschema.yml
Comment (GernotMaier) [reply simtools/schemas/model_parameter_and_data_schema.metaschema.yml:102]: No, we always had draft 6 (the format checker for astropy units was the only place where we mentioned Draft7.

I honestly don't know enough about the differences.

**Comment by GernotMaier** [comment]
@orelgueta - thanks for the review. I've improved the unit tests for format checker.

Let me know if you insist to add a test for this one line in `validate_data`. I would then refactor the complete function to make it cleaner.

File: simtools/data_model/format_checkers.py
Comment (orelgueta) [reply simtools/data_model/format_checkers.py:22]: Yeah I saw later and didn't delete this comment.

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (orelgueta) [reply tests/unit_tests/data_model/test_format_checkers.py:10]: Thanks!

File: simtools/schemas/model_parameter_and_data_schema.metaschema.yml
Comment (orelgueta) [reply simtools/schemas/model_parameter_and_data_schema.metaschema.yml:102]: OK, before I saw something about Draft7, but it doesn't matter.

Comment (GernotMaier) [reply simtools/schemas/model_parameter_and_data_schema.metaschema.yml:102]: It was here: https://github.com/gammasim/simtools/blob/1c3acd9c1e7498214527db7d3045aa2447804e5d/simtools/data_model/metadata_model.py#L24

But this line has been changed now that it is independent of the draft. Really not sure why we didn't do it like that from the beginning.

Comment (orelgueta) [reply simtools/schemas/model_parameter_and_data_schema.metaschema.yml:102]: OK!

File: tests/unit_tests/data_model/test_format_checkers.py
Comment (orelgueta) [reply tests/unit_tests/data_model/test_format_checkers.py:22]: Cause I mixed up this and the other function which I thought it calls `_validate_data_dict_using_json_schema`. I pulled out the branch and looked at the code more carefully now and this comment is simply wrong.

File: simtools/data_model/validate_data.py
Comment (orelgueta) [reply simtools/data_model/validate_data.py:147]: I am really hesitant to be honest. I agree that the line itself is not important cause the function it calls is tested elsewhere.  On the other hand, it definitely points to the fact that this function would be better if it were simplified. I leave it to you to decide then. If you think refactoring is worth it, good. Otherwise I approve already and we can ignore this warniing.

**Review by orelgueta — APPROVED** [review]
Approving with the caveat that refactoring would be useful, but not necessary.

File: simtools/data_model/validate_data.py
Comment (GernotMaier) [reply simtools/data_model/validate_data.py:147]: Will refactoring it, but not as part of this pull request: #1193

**Comment by GernotMaier** [comment]
@orelgueta - thanks again. I moved the remaining point to an issue and will merge this.

**Pull request merged**
